nntrainer/tensor/tensor.cpp

```diff
@@ -445,12 +451,20 @@ Tensor Tensor::sum(const std::vector<unsigned int> &axes, float alpha) const {
   return ret;
 }
 
+Tensor Tensor::dot(Tensor const &m, bool trans, bool trans_m) const {
+  Tensor output;
+  dot(m, output, trans, trans_m);
+
+  return output;
+}
+
 /**
  * @note: This dot product flattens the fist 3 axis for the purpose of
  * computation. So, while performing, these matrices are behaving as 2-D
  * matrices. The dimensions are restored while returning back the tensor.
  */
-Tensor Tensor::dot(Tensor const &m, bool trans, bool trans_m) const {
+Tensor Tensor::dot(Tensor const &m, Tensor &result, bool trans,
+                   bool trans_m) const {
```

Comment:
How about we make sure  `&m.getData() = &result.getData()`? Eventhough m is const, underlying data can change in this case.

Reply:
I guess this will prevail, so instead putting all checks in every operation, I wish there is a better way to do

Reply:
`&result != this` is necessary in this case as well

Reply:
you mean inverse of `&m.getData() = &result.getData()`?? (that they should not be same?)
but this wont be sufficient, it will have to be `&m.getData() + m.length()` and `&result.getData() + result.length()` (if result is preallocated) are not intersecting.

Reply:
sorry, I meant sth like,

```cpp
if(m.getData() == result.getData() || this->getData() == result.getData()) {
 throw
} 
```

As you commented out, I guess checking if intersecting should be done

Reply:
These overlapping checks and return type updates will do in next PR if that's ok, as they will have to be done for a class of operations.

Reply:
No worries